lib/api_callable.js

```diff
@@ -152,137 +220,170 @@ function retryable(aFunc, retry, otherArgs) {
-    return function fetchNextPageToken(err, response) {
-      if (err) {
-        callback(err);
-      }
-      callback(err, response, response[pageDescriptor.responsePageTokenField]);
-    };
+function PageStreamable(pageDescriptor) {
+  this.pageDescriptor = pageDescriptor;
+}
+
+PageStreamable.prototype.createActualCallback = function(callback) {
+  return function fetchNextPageToken(err, response) {
+    if (err) {
+      callback(err);
+    }
+    var pageToken = response[this.pageDescriptor.responsePageTokenField];
+    callback(err, response, pageToken);
+  }.bind(this);
+};
+
+PageStreamable.prototype.init = function(settings, callback) {
+  if (callback) {
+    settings.flattenPages = false;
+    callback = this.createActualCallback(callback);
   }
-  return {
-    init: function createStream(settings, callback) {
-      if (callback) {
-        settings.flattenPages = false;
-        callback = createActualCallback(callback);
-      }
-      if (!settings.flattenPages) {
-        return new EventEmitter(callback);
-      }
-      return through2.obj();
-    },
-    call: function pageStreaming(apiCall, argument, settings, stream) {
-      if (settings.pageToken) {
-        argument[pageDescriptor.requestPageTokenField] = settings.pageToken;
-      }
-      if (!settings.flattenPages) {
-        // stream is an emitter, calling as a normal API call.
-        stream._setEventHandler(apiCall(argument, stream._callback));
-        return;
-      }
+  if (!settings.flattenPages) {
+    return NormalApiCaller.prototype.init.call(this, settings, callback);
+  }
+  this.stream = through2.obj();
+  return this.stream;
+};
 
-      /** Execute apiCall as long as necessary or to the end of the stream. */
-      function streaming() {
-        apiCall(argument, function(err, response) {
-          if (err) {
-            stream.emit('error', err);
-            stream.end();
-            return;
-          }
-          if (!stream.writable) {
-            return;
-          }
-          var resources = response[pageDescriptor.resourceField];
-          for (var i in resources) {
-            if (!stream.writable) {
-              break;
-            }
-            stream.write(resources[i]);
-          }
-          if (!stream.writable) {
-            return;
-          }
-          var nextPageToken = response[pageDescriptor.responsePageTokenField];
-          if (!nextPageToken) {
-            stream.end();
-            return;
-          }
-          argument[pageDescriptor.requestPageTokenField] = nextPageToken;
-          setTimeout(streaming, 0);
-        });
-      }
-      setTimeout(streaming, 0);
+PageStreamable.prototype.streamingCallback = function(err, response) {
+  if (err) {
+    this.stream.emit('error', err);
+    this.stream.end();
+    return;
+  }
+  if (!this.stream.writable) {
+    return;
+  }
+
+  var resources = response[this.pageDescriptor.resourceField];
+  for (var i in resources) {
+    if (!this.stream.writable) {
```

Comment:
'data' event happens on each `write` call. Think that the user only cares about the first element in the stream, then
```javascript
var s = api.pageStreamingMethod();
s.on('data', function(data) {
  doSomething(data);
  s.end();
});
```

In that case, `stream.writable` turns to false for the first invocation of `stream.write`.

Or, see the following code:
```javascript
var s = through2.obj();
s.on('data', function(d) { console.log(d); s.end(); });
console.log(s.writable); // true
s.write('foo'); // print out 'foo'
console.log(s.writable); // false, because on-data callback is executed.
s.write('foo'); // causes an exception because it's not writable.
```

That means, however, the section around line 310 wouldn't be necessary (line 321 is still necessary, in case `stream.writable` turns to false at the last element).

Also updates the test case (see 'stops in the middle') to clarify the condition.